### DISCUSSION TITLE
Using the webclient module in a program causes Berry to crash

### DISCUSSION THREAD

**Discussion by DremOSDeveloperTeam** [arendst/Tasmota#18832] [Q&A] [answered]
Greetings

First off, sorry I'm in the wrong place. I'm aware of the [Berry repo](https://github.com/berry-lang/berry/), but I'm reasonably confident this is an issue for here. If not, please kindly direct me to the right place.

That out of the way...

# The Problem

When I use the webclient module, `cl.GET()` causes the program to stop, and I assume Berry crashes as well (or maybe gets stuck in a loop?) because REPL will stop responding. I'm able to send text to it, but it never says anything back.

The strange thing is, however, that running code similar to what I'll paste below in the REPL works fine. For example:

```
var cl = webclient()
cl.begin("http://192.168.4.221/cm?cmnd=POWER+STATUS")
print(cl.GET())
print(cl.get_string())
```
Output:
```
200
{"POWER":"OFF"}
```

# My Code

Here is my code. 

The local address is a different Tasmota device.

```
class TestClient
    def init()
        print("Test client start")
    end

    def every_second()
        var cl = webclient()
        cl.begin("http://192.168.4.221/cm?cmnd=POWER+STATUS")
        print(cl.GET()) # The problem. This is what causes the crash.
        print(cl.get_string())
    end
end

var testclient = module("TestClient")
testclient.TestClient = TestClient()

return TestClient
```

# My Device

My device is a [Ulanzi Smart Pixel Clock (TC001)](https://templates.blakadder.com/ulanzi_TC001.html), running Tasmota 12.5.0.

**Comment by s-hadinger** [comment, marked as answer]
Currently, if the Wifi is not yet established, webclient may crash. This is not ideal but we hadn't had time to fix this.

One way to guard agains it is:
```berry
if !tasmota.wifi()['up'] && !tasmota.eth()['up']    return   end 
```

**Reply by DremOSDeveloperTeam** [reply]
Ah, I see. Thank you very much!

**Comment by sfromis** [comment]
I'd strongly recommend closing the cl instance of webclient, instead of relying on garbage collection to clean up later. Not doing this may increase the risk of crashing.

Also, I'd like add a couple of things outside of the crash issue.

The command `POWER STATUS` is invalid. Not sure what you expect, but it should still return something.

For monitoring status of another device, it is highly recommended to not use polling, but instead subscribe to MQTT messages already published by the other device when status changes.

Some devices may not be able to sustain being polled every second, but crash due to that. To be clear, this would not be Berry crashing.

**Reply by DremOSDeveloperTeam** [reply]
I read in the docs that `webclient.get_string` also closes the webclient, but I should probably close it myself.

As for `POWER STATUS`, it works fine for me. Is there a different command I'm supposed to use?

I haven't yet looked into MQTT, but I'll look into that soon™. In my experience, polling fairly often has not yet come up as an issue, but if MQTT would use up less resources it would obviously be preferred. Thanks for this suggestion!

**Reply by sfromis** [reply]
Well, the command **is** invalid. If you want to merely have power status, the command is just "Power", nothing more.

Frequent polling causing issues is a common problem with the TasmoAdmin tool, where many users report the device much more stable without that load. Using MQTT will have a hugely (in relative terms, of course) lower resource cost, as the other device will publish when status changes, instead of the wasteful polling process, which is still slower than relying on MQTT messages.

Close - I had not seen the note about it freeing resources. As an old time computer user, I may be a bit over-sensitive to seeing code not using available cleanup calls :grin:

**Reply by barbudor** [reply]
> I may be a bit over-sensitive to seeing code not using available cleanup calls 😁

👍 

![image](https://github.com/arendst/Tasmota/assets/2996042/355f6167-d384-4acf-b2a3-e66fc1865fb1)

**Reply by DremOSDeveloperTeam** [reply]
> the command is just "Power", nothing more.

Ah I see. I just had like a realization that it just ignores what you put after it if it's invalid while messing around with MQTT just now (a result of making a typo with the payload, and seeing something show up on the stat topic).

> Using MQTT will have a hugely (in relative terms, of course) lower resource cost

I noticed that immediately. Like a weight being removed from my back, as I stop having to wait like 2 seconds for the API to respond.

> I may be a bit over-sensitive to seeing code not using available cleanup calls grin

Ay, I was a compsci student, I get that. No leaks on my watch!

**Comment by ghost** [comment]
I'm currently trying something similar. The example in the documentation works. When I change the link to Google https://script.google.com, Berry seems to crash. In any case, I get no feedback and there is no entry in the Google Sheet either.

Have you found a solution?

My code:

```
cl = webclient()
cl.set_follow_redirects(true)
cl.begin("https://script.google.com/macros/s/--App-Script-ID--/exec?D=20.04.2025,17:00:01&I=123&E=456&G=789")
r = cl.GET()
print(r)
s = cl.get_string()
print(s)
```

**Reply by sfromis** [reply]
Trying with that code, I did not see any crashing, but instead outputs "1" for the first print of status code, and a blank line for the content string.

Testing same URL in a browser, I'm seeing http status 500, internal server error. Seems not to be a usable URL.

**Reply by s-hadinger** [reply]
One possibility is that the response is too big to fit in memory. You can read chunks instead

**Reply by ghost** [reply]
Sorry, my mistake. Everything works now. The script was not released for everyone (without Google account).
